src/main/scala/strawman/collection/immutable/HashSet.scala

```diff
@@ -2,40 +2,325 @@ package strawman
+/** This class implements immutable sets using a hash trie.
+  *
+  *  '''Note:''' The builder of this hash set may return specialized representations for small sets.
+  *
+  *  @tparam A      the type of the elements contained in this hash set.
+  *
+  *  @author  Martin Odersky
+  *  @author  Tiark Rompf
+  *  @version 2.8
+  *  @since   2.3
+  *  @define Coll `immutable.HashSet`
+  *  @define coll immutable hash set
+  */
+@SerialVersionUID(2L)
+sealed trait HashSet[A]
+  extends Set[A]
+    with SetLike[A, HashSet]
+    with Serializable {
 
-  // From IterableOnce
-  def iterator(): Iterator[A] = ???
+  import HashSet.nullToEmpty
 
   // From IterablePolyTransforms
   def fromIterable[B](coll: collection.Iterable[B]): HashSet[B] = HashSet.fromIterable(coll)
   protected[this] def fromIterableWithSameElemType(coll: collection.Iterable[A]): HashSet[A] = fromIterable(coll)
 
   // From SetLike
-  def contains(elem: A): Boolean = ???
-  def subsetOf(that: collection.Set[A]): Boolean = ???
+  def contains(elem: A): Boolean = get0(elem, computeHash(elem), 0)
 
-  // From SetMonoTransforms
-  def & (that: collection.Set[A]): HashSet[A] = ???
-  def ++ (that: collection.Set[A]): HashSet[A] = ???
+  def + (elem: A): HashSet[A] = updated0(elem, computeHash(elem), 0)
 
-  // From immutable.SetLike
-  def + (elem: A): HashSet[A] = ???
-  def - (elem: A): HashSet[A] = ???
+  def - (elem: A): HashSet[A] = nullToEmpty(removed0(elem, computeHash(elem), 0))
+
+  protected def get0(key: A, hash: Int, level: Int): Boolean
+
+  protected def updated0(key: A, hash: Int, level: Int): HashSet[A]
+
+  protected def removed0(key: A, hash: Int, level: Int): HashSet[A]
+
+  protected def computeHash(key: A): Int = improve(elemHashCode(key))
+
+  protected final def improve(hcode: Int): Int = {
+    var h: Int = hcode + ~(hcode << 9)
+    h = h ^ (h >>> 14)
+    h = h + (h << 4)
+    h ^ (h >>> 10)
+  }
+
+  protected final def elemHashCode(elem: A): Int = elem.##
 
 }
 
 object HashSet extends IterableFactory[HashSet] {
 
-  def fromIterable[A](it: collection.Iterable[A]): HashSet[A] = ???
+  def fromIterable[A](it: collection.Iterable[A]): HashSet[A] = newBuilder[A].++=(it).result
+
+  def newBuilder[A]: Builder[A, HashSet[A]] = new ImmutableSetBuilder[A, HashSet](empty[A])
+
+  override def empty[A <: Any]: HashSet[A] = EmptyHashSet.asInstanceOf[HashSet[A]]
+
+  private object EmptyHashSet extends HashSet[Any] {
+
+    def iterator(): Iterator[Any] = Iterator.empty
+
+    override def foreach[U](f: Any => U): Unit = ()
+
+    override def head: Any = throw new NoSuchElementException("Empty Set")
+
+    override def tail: HashSet[Any] = throw new NoSuchElementException("Empty Set")
+
+    override def size: Int = 0
+
+    protected def get0(elem: Any, hash: Int, level: Int) = false
+
+    protected def updated0(elem: Any, hash: Int, level: Int) = new HashSet1(elem, hash)
+
+    protected def removed0(key: Any, hash: Int, level: Int) = this
+
+  }
```

Comment:
> I don’t know if it has an impact on performance

I wouldn't expect a measurable difference, but it's better to validate using benchmarks.